File: examples/java/src/main/java/SimpleExample.java
import io.triplox.client.Db;
import io.triplox.client.TriploxNode;
import io.triplox.client.TxOp;
import io.triplox.client.TxResult;

import java.util.List;
import java.util.Map;

import static io.triplox.client.Util.kw;
import static io.triplox.client.Util.map;

public final class SimpleExample {
    private SimpleExample() {
    }

    private static Map<String, Object> schemaAttribute(String name, String valueType) {
        return map(
                ":db/ident", kw(":" + name),
                ":db/valueType", kw(":db.type/" + valueType),
                ":db/cardinality", kw(":db.cardinality/one"));
    }

    private static TxResult requireCommitted(String label, TxResult result) {
        if (!result.isCommitted()) {
            throw new IllegalStateException(label + " transaction aborted: " + result.errorMessage());
        }
        return result;
    }

    public static void main(String[] args) throws Exception {
        var host = "localhost";
        var port = 5490;
        System.out.println("Connecting to " + host + ":" + port + "...");

        try (var node = TriploxNode.connect(host, port)) {
            System.out.println("Connected.");

            var schemaResult = requireCommitted("Schema", node.executeTx(List.of(
                    new TxOp.Put(schemaAttribute("name", "string")),
                    new TxOp.Put(schemaAttribute("age", "long")))));
            System.out.println("Schema defined (tx_id=" + schemaResult.txId() + ").");

            var dataResult = requireCommitted("Data", node.executeTx(List.of(
                    new TxOp.Put(map(":name", "alice", ":age", 30L)),
                    new TxOp.Put(map(":name", "bob", ":age", 25L)))));
            System.out.println("Data inserted (tx_id=" + dataResult.txId() + ").");

            try (Db db = node.openDb()) {
                System.out.println("Opened DB snapshot (tx_eid=" + db.txEid() + ").");
                var rows = db.query("{:find [?e ?name ?age] :where [[?e :name ?name] [?e :age ?age]]}");

                System.out.println("Query returned " + rows.size() + " row(s):");
                for (var row : rows) {
                    System.out.println("  " + row);
                }
            }
        }

        System.out.println("Done.");
    }
}

File: examples/rust/src/simple_example.rs
//! Simple example: connect to a running Triplox server, define a schema
//! attribute, insert data, and query it back.
//!
//! Start the server first:
//!   cargo run                     (from project root, uses config/triplox.toml)
//!
//! Then run this example:
//!   cargo run --bin simple-example  (from examples/rust/)

use anyhow::Result;
use edn::kw;
use edn::Keyword;
use triplox::client::ClientNode;
use triplox::node::{Database, QueryNode, SubmitNode, TransactionResult};
use triplox::ops::{DataType, TxOp};

/// Build a schema attribute definition as a Put document.
/// This mirrors the internal `plain_schema_attribute` helper.
fn schema_attribute(name: &str, value_type: &str) -> TxOp {
    TxOp::put(vec![
        (kw!(:db/ident), DataType::Keyword(Keyword::plain(name))),
        (kw!(:db/valueType), DataType::Keyword(Keyword::namespaced("db.type", value_type))),
        (kw!(:db/cardinality), DataType::Keyword(kw!(:db.cardinality/one))),
    ])
}

#[tokio::main]
async fn main() -> Result<()> {
    let addr = "http://127.0.0.1:5490";
    println!("Connecting to {addr}...");
    let node = ClientNode::connect(addr).await?;
    println!("Connected.");

    // 1. Define schema attributes
    let schema_ops = vec![
        schema_attribute("name", "string"),
        schema_attribute("age", "long"),
    ];
    let result = node.execute_tx(schema_ops).await?;
    match &result {
        TransactionResult::TxCommited(tx_key) => {
            println!("Schema defined (tx_id={}).", tx_key.tx_id);
        }
        TransactionResult::TxAborted(_, err) => {
            anyhow::bail!("Schema transaction aborted: {err}");
        }
    }

    // 2. Insert some data
    let data_ops = vec![
        TxOp::put(vec![
            (kw!(:name), "alice".into()),
            (kw!(:age), 30_i64.into()),
        ]),
        TxOp::put(vec![
            (kw!(:name), "bob".into()),
            (kw!(:age), 25_i64.into()),
        ]),
    ];
    let result = node.execute_tx(data_ops).await?;
    match &result {
        TransactionResult::TxCommited(tx_key) => {
            println!("Data inserted (tx_id={}).", tx_key.tx_id);
        }
        TransactionResult::TxAborted(_, err) => {
            anyhow::bail!("Data transaction aborted: {err}");
        }
    }

    // 3. Open a DB value and query
    let db = node.db().await?;
    println!("Opened DB value (tx_eid={}).", db.tx_eid());

    let rows = db
        .query(r#"{:find [?e ?name ?age] :where [[?e :name ?name] [?e :age ?age]]}"#)
        .await?;

    println!("Query returned {} row(s):", rows.len());
    for row in &rows {
        println!("  {:?}", row);
    }

    println!("Done.");

    Ok(())
}
